### ISSUE TITLE
Performance hit in OgRoleCacheContext

### ISSUE THREAD

**Issue by pfrenssen** [Gizra/og#487]
`OgRoleCacheContext::getContext()` is slow for users that are a member of a large number of groups. This is critical since this code is running in order to calculate the cache IDs of items that vary by their group roles. This means that even on pages that are fully cached, this code needs to run in order to retrieve the correct items from the page cache or render cache. 

For users that are a member of hundreds or thousands of groups this is causing cached pages to load slowly.

Here is the code in question:

```
      foreach ($this->membershipManager->getMemberships($this->user) as $membership) {
        $role_names = array_map(function (OgRoleInterface $role) {
          return $role->getName();
        }, $membership->getRoles());
        if ($role_names) {
          $memberships[$membership->getGroupEntityType()][$membership->getGroupId()] = $role_names;
        }
      }
```

This code is doing multiple slow operations:

1. It loads the full `OgMembership` entities of all the user's memberships. This is a lot of data.
2. For each membership it calls `$membership->getRoles()` which returns the full `OgRole` entities for the membership. These are a lot of individual queries.

I investigated the code path and there are a number of improvements we can make to make this a lot faster, and also improve some parts that will make the OG API faster for our users:

- [ ] `OgMembership::getMemberships()` can be persistently cached. See #485.
- [x] `OgMembership::getRolesIds()` does not need to call `$this->getRoles()` in order to get the role IDs. These IDs are already present on the entity reference field. See #555.
- [x] Once the previous step is done, `OgRoleCacheContext::getContext()` should call `$membership->getRolesIds()` instead of `::getRoles()`. We can then get the role name from the ID, or even use the full ID. See #555.

This still leaves us with the problem that all the `OgMembership` entities are loaded. We could solve this by doing a `SELECT` query on the database that directly returns the data from the role field. This would be very fast, but would make OG incompatible with NoSQL databases like MongoDB / CouchDB / Virtuoso. This can still be solved by checking if the `OgMembership` database backend implements `SqlContentEntityStorage` and falling back to the inefficient loading of all memberships if needed.

**Comment by amitaibu** [comment]
> but would make OG incompatible with NoSQL databases like MongoDB / CouchDB / Virtuoso
> This can still be solved by checking if the OgMembership database backend implements SqlContentEntityStorage and falling back to the inefficient loading of all memberships if needed.

I believe NoSQL databases with OG is an edge case - one I have never heard in real life. and I'd prefer to have quicker OG for 99% (of not 100%) then say we support NoSQL databases

**Comment by pfrenssen** [comment]
I agree that we should focus on the 99%. Still, if we can provide a (slower) fallback that still supports NoSQL then it's even better.

FYI we are using OG with a NoSQL database so you can rest assured that it works great out of the box :) We are using a [SPARQL](https://www.w3.org/TR/sparql11-query/) database which offers all our group data as open data on the web. We are developing a module to support SPARQL in Drupal: https://www.drupal.org/project/rdf_entity

We are not publishing our group membership information as open data though, so our `OgMembership` entities are still in a standard SQL database.

**Comment by amitaibu** [comment]
> FYI we are using OG with a NoSQL database

Wow, That's a first I hear, of someone actually using it :smile: . In that case, yeah - we could fallback upon checking `SqlContentEntityStorage`

**Comment by pfrenssen** [comment]
I have completed the work on this in 2 different PRs. I have done profiling at each step.

The results of the testing has made me decide to not pursue the persistent caching of the `MembershipManager` static cache since it would have very little effect (estimated to only ~1-2 ms of speed gain).

The profiling has indicated another unexpected performance bottleneck: retrieving field data appears to be very slow, almost as slow as retrieving data from the database cache. This is also fixed in #555.

Here are the profiling results. The tests were done using a user with 1142 memberships, on a warm cache with assertions turned off. The time shown is the run through time of `OgRoleCacheContext::getContext()`, averaged over 5 requests.

Time | Description
--- | ---
154.86ms | Baseline before optimizing
122.71ms | Switch from `$membership->getRoles()` to `$membership->getRolesIds()` (#555 - first part)
73.61ms | Speed up field reads (#555 - second part)
11.13ms | Use SELECT query on SQL databases (#559)

So the final result is that for SQL databases the cache context is now 13.9x faster, and the fallback for NoSQL databases is ~2.1x faster.

**Comment by geek-merlin** [comment]
Without having fully dived into the details... Did you notice [this issue](https://www.drupal.org/project/drupal/issues/2551419) where big improvements may just be happening for big-cardinality cache contexts.

**Comment by pfrenssen** [comment]
@geek-merlin thanks for pointing me to that issue, it looks very interesting. I subscribed to it.

It looks like this will provide even more benefits, since it will allow us to cache the result inbetween requests. It is a huge issue though and I just spent a few minutes reading.

Even with that issue we will still need this, since this improves the performance of doing the actual generation of the cache context key. This cannot be avoided, even though it would be possible to cache the result of it.
